File: python/src/traffic_control/solvers/max_flow.py
from __future__ import annotations

from collections import deque
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from traffic_control.models import Network, FlowSolution, SolverMethod


class MaxFlowError(Exception):
    pass


def edmonds_karp(network: "Network", source: str, sink: str) -> "FlowSolution":
    from traffic_control.models import FlowSolution, SolverMethod

    road_to_idx = {r.id: i for i, r in enumerate(network.roads)}
    n = len(network.junctions)
    junction_to_idx = {j.id: i for i, j in enumerate(network.junctions)}

    cap = [[0] * n for _ in range(n)]
    road_map = [[None] * n for _ in range(n)]

    for road in network.roads:
        u = junction_to_idx[road.source]
        v = junction_to_idx[road.target]
        cap[u][v] += road.capacity
        road_map[u][v] = road.id

    s = junction_to_idx.get(source)
    t = junction_to_idx.get(sink)

    if s is None or t is None:
        raise MaxFlowError(f"Source '{source}' or sink '{sink}' not found")

    flow = [[0] * n for _ in range(n)]
    max_flow = 0

    while True:
        parent = [-1] * n
        parent[s] = s
        q = deque([s])
        path_cap = [0] * n
        path_cap[s] = float("inf")

        while q and parent[t] == -1:
            u = q.popleft()
            for v in range(n):
                if parent[v] == -1 and cap[u][v] - flow[u][v] > 1e-9:
                    parent[v] = u
                    path_cap[v] = min(path_cap[u], cap[u][v] - flow[u][v])
                    q.append(v)

        if parent[t] == -1:
            break

        path_flow = path_cap[t]
        v = t
        while v != s:
            u = parent[v]
            flow[u][v] += path_flow
            flow[v][u] -= path_flow
            v = u

        max_flow += path_flow

    flows = {r.id: 0.0 for r in network.roads}
    for u in range(n):
        for v in range(n):
            if flow[u][v] > 1e-9 and road_map[u][v]:
                flows[road_map[u][v]] = flow[u][v]

    return FlowSolution(
        flows=flows,
        method=SolverMethod.MAX_FLOW,
        objective_value=max_flow,
        is_feasible=True,
        metadata={
            "algorithm": "edmonds_karp",
            "source": source,
            "sink": sink,
            "max_flow": max_flow,
        }
    )


def dinic(network: "Network", source: str, sink: str) -> "FlowSolution":
    from traffic_control.models import FlowSolution, SolverMethod

    n = len(network.junctions)
    junction_to_idx = {j.id: i for i, j in enumerate(network.junctions)}
    road_to_idx = {r.id: i for i, r in enumerate(network.roads)}

    s = junction_to_idx.get(source)
    t = junction_to_idx.get(sink)

    if s is None or t is None:
        raise MaxFlowError(f"Source '{source}' or sink '{sink}' not found")

    adj = [[] for _ in range(n)]
    road_ids = {}

    for road in network.roads:
        u = junction_to_idx[road.source]
        v = junction_to_idx[road.target]
        adj[u].append([v, len(adj[v]), road.capacity, road.id])
        adj[v].append([u, len(adj[u]) - 1, 0.0, None])
        road_ids[(u, len(adj[u]) - 1)] = road.id

    def bfs() -> list[int]:
        level = [-1] * n
        level[s] = 0
        q = deque([s])
        while q:
            u = q.popleft()
            for v, rev, cap, _ in adj[u]:
                if cap > 1e-9 and level[v] == -1:
                    level[v] = level[u] + 1
                    q.append(v)
        return level

    def dfs(u: int, flow: float, level: list[int], it: list[int]) -> float:
        if u == t:
            return flow
        for i in range(it[u], len(adj[u])):
            it[u] = i
            v, rev, cap, rid = adj[u][i]
            if cap > 1e-9 and level[v] == level[u] + 1:
                pushed = dfs(v, min(flow, cap), level, it)
                if pushed > 1e-9:
                    adj[u][i][2] -= pushed
                    adj[v][rev][2] += pushed
                    return pushed
        return 0.0

    max_flow = 0.0
    while True:
        level = bfs()
        if level[t] == -1:
            break
        it = [0] * n
        while True:
            pushed = dfs(s, float("inf"), level, it)
            if pushed < 1e-9:
                break
            max_flow += pushed

    flows = {r.id: 0.0 for r in network.roads}
    for u in range(n):
        for idx, (v, rev, cap, rid) in enumerate(adj[u]):
            if rid is not None:
                original_cap = sum(r.capacity for r in network.roads if r.id == rid)
                flow_val = original_cap - cap
                if flow_val > 1e-9:
                    flows[rid] = flow_val

    return FlowSolution(
        flows=flows,
        method=SolverMethod.DINIC,
        objective_value=max_flow,
        is_feasible=True,
        metadata={
            "algorithm": "dinic",
            "source": source,
            "sink": sink,
            "max_flow": max_flow,
        }
    )


def solve_max_flow(
    network: "Network",
    source: str,
    sink: str,
    algorithm: str = "dinic"
) -> "FlowSolution":
    if algorithm == "edmonds_karp":
        return edmonds_karp(network, source, sink)
    elif algorithm == "dinic":
        return dinic(network, source, sink)
    else:
        raise MaxFlowError(f"Unknown algorithm: {algorithm}")

File: rust/src/solvers/max_flow.rs
use petgraph::graph::{DiGraph, NodeIndex};
use petgraph::visit::EdgeRef;
use std::collections::{HashMap, VecDeque};
use crate::models::{Network, FlowSolution, SolverMethod};
use crate::network::to_petgraph;

#[derive(Debug, thiserror::Error)]
pub enum MaxFlowError {
    #[error("Source or sink not found")]
    NotFound,
    #[error("No path from source to sink")]
    NoPath,
}

pub fn dinic(network: &Network, source: &str, sink: &str) -> Result<FlowSolution, MaxFlowError> {
    let graph = to_petgraph(network);
    let node_map: HashMap<_, _> = graph.node_indices()
        .map(|n| (graph[n].id.clone(), n))
        .collect();

    let src = node_map.get(source).ok_or(MaxFlowError::NotFound)?;
    let sink = node_map.get(sink).ok_or(MaxFlowError::NotFound)?;

    let mut cap = HashMap::new();
    let mut flow = HashMap::new();
    let mut adj = HashMap::new();

    for edge in graph.edge_references() {
        let u = edge.source();
        let v = edge.target();
        let road = edge.weight();
        cap.insert((u, v), road.capacity);
        cap.insert((v, u), 0.0);
        flow.insert((u, v), 0.0);
        flow.insert((v, u), 0.0);

        adj.entry(u).or_insert_with(Vec::new).push(v);
        adj.entry(v).or_insert_with(Vec::new).push(u);
    }

    let mut max_flow = 0.0;

    while bfs_level(&graph, &cap, &flow, src, sink, &mut HashMap::new()) {
        let mut level = HashMap::new();
        bfs_level(&graph, &cap, &flow, src, sink, &mut level);
        let mut it = HashMap::new();

        while let Some(pushed) = dfs_blocking(&graph, &cap, &mut flow, &adj, &level, &mut it, *src, *sink, f64::INFINITY) {
            max_flow += pushed;
        }
    }

    let mut flows = HashMap::new();
    for edge in graph.edge_references() {
        let u = edge.source();
        let v = edge.target();
        let road = edge.weight();
        let f = flow.get(&(u, v)).copied().unwrap_or(0.0);
        if f > 1e-9 {
            flows.insert(road.id.clone(), f);
        }
    }

    for road in &network.roads {
        flows.entry(road.id.clone()).or_insert(0.0);
    }

    Ok(FlowSolution {
        flows,
        method: SolverMethod::Dinic,
        objective: None,
        objective_value: Some(max_flow),
        is_feasible: true,
        violations: Vec::new(),
        metadata: {
            let mut m = HashMap::new();
            m.insert("maxFlow".to_string(), serde_json::Value::Number(serde_json::Number::from_f64(max_flow).unwrap()));
            m.insert("source".to_string(), serde_json::Value::String(source.to_string()));
            m.insert("sink".to_string(), serde_json::Value::String(sink.to_string()));
            m
        },
    })
}

fn bfs_level(
    graph: &DiGraph<crate::models::Junction, crate::models::Road>,
    cap: &HashMap<(NodeIndex, NodeIndex), f64>,
    flow: &HashMap<(NodeIndex, NodeIndex), f64>,
    src: &NodeIndex,
    sink: &NodeIndex,
    level: &mut HashMap<NodeIndex, i32>,
) -> bool {
    level.clear();
    let mut q = VecDeque::new();
    level.insert(*src, 0);
    q.push_back(*src);

    while let Some(u) = q.pop_front() {
        for edge in graph.edges(u) {
            let v = edge.target();
            let residual = cap.get(&(u, v)).copied().unwrap_or(0.0) - flow.get(&(u, v)).copied().unwrap_or(0.0);
            if residual > 1e-9 && !level.contains_key(&v) {
                level.insert(v, level[&u] + 1);
                q.push_back(v);
            }
        }
    }

    level.contains_key(sink)
}

fn dfs_blocking(
    graph: &DiGraph<crate::models::Junction, crate::models::Road>,
    cap: &HashMap<(NodeIndex, NodeIndex), f64>,
    flow: &mut HashMap<(NodeIndex, NodeIndex), f64>,
    adj: &HashMap<NodeIndex, Vec<NodeIndex>>,
    level: &HashMap<NodeIndex, i32>,
    it: &mut HashMap<NodeIndex, usize>,
    u: NodeIndex,
    sink: NodeIndex,
    f: f64,
) -> Option<f64> {
    if u == sink {
        return Some(f);
    }

    let neighbors = adj.get(&u)?;
    let start = *it.get(&u).unwrap_or(&0);

    for i in start..neighbors.len() {
        let v = neighbors[i];
        it.insert(u, i + 1);

        let residual = cap.get(&(u, v)).copied().unwrap_or(0.0) - flow.get(&(u, v)).copied().unwrap_or(0.0);
        if residual > 1e-9 && level.get(&v) == Some(level[&u] + 1) {
            if let Some(pushed) = dfs_blocking(graph, cap, flow, adj, level, it, v, sink, f.min(residual)) {
                *flow.get_mut(&(u, v)).unwrap() += pushed;
                *flow.get_mut(&(v, u)).unwrap() -= pushed;
                return Some(pushed);
            }
        }
    }

    None
}

pub fn edmonds_karp(network: &Network, source: &str, sink: &str) -> Result<FlowSolution, MaxFlowError> {
    let graph = to_petgraph(network);
    let node_map: HashMap<_, _> = graph.node_indices()
        .map(|n| (graph[n].id.clone(), n))
        .collect();

    let src = node_map.get(source).ok_or(MaxFlowError::NotFound)?;
    let sink = node_map.get(sink).ok_or(MaxFlowError::NotFound)?;

    let mut cap = HashMap::new();
    let mut flow = HashMap::new();
    let mut adj = HashMap::new();

    for edge in graph.edge_references() {
        let u = edge.source();
        let v = edge.target();
        let road = edge.weight();
        cap.insert((u, v), road.capacity);
        cap.insert((v, u), 0.0);
        flow.insert((u, v), 0.0);
        flow.insert((v, u), 0.0);
        adj.entry(u).or_insert_with(Vec::new).push(v);
        adj.entry(v).or_insert_with(Vec::new).push(u);
    }

    let mut max_flow = 0.0;

    while let Some((path_flow, parent)) = bfs_augmenting_path(&graph, &cap, &flow, &adj, *src, *sink) {
        let mut v = *sink;
        while v != *src {
            let u = parent[&v];
            *flow.get_mut(&(u, v)).unwrap() += path_flow;
            *flow.get_mut(&(v, u)).unwrap() -= path_flow;
            v = u;
        }
        max_flow += path_flow;
    }

    let mut flows = HashMap::new();
    for edge in graph.edge_references() {
        let u = edge.source();
        let v = edge.target();
        let road = edge.weight();
        let f = flow.get(&(u, v)).copied().unwrap_or(0.0);
        if f > 1e-9 {
            flows.insert(road.id.clone(), f);
        }
    }

    for road in &network.roads {
        flows.entry(road.id.clone()).or_insert(0.0);
    }

    Ok(FlowSolution {
        flows,
        method: SolverMethod::MaxFlow,
        objective: None,
        objective_value: Some(max_flow),
        is_feasible: true,
        violations: Vec::new(),
        metadata: {
            let mut m = HashMap::new();
            m.insert("maxFlow".to_string(), serde_json::Value::Number(serde_json::Number::from_f64(max_flow).unwrap()));
            m.insert("algorithm".to_string(), serde_json::Value::String("edmonds_karp".to_string()));
            m.insert("source".to_string(), serde_json::Value::String(source.to_string()));
            m.insert("sink".to_string(), serde_json::Value::String(sink.to_string()));
            m
        },
    })
}

fn bfs_augmenting_path(
    graph: &DiGraph<crate::models::Junction, crate::models::Road>,
    cap: &HashMap<(NodeIndex, NodeIndex), f64>,
    flow: &HashMap<(NodeIndex, NodeIndex), f64>,
    adj: &HashMap<NodeIndex, Vec<NodeIndex>>,
    src: NodeIndex,
    sink: NodeIndex,
) -> Option<(f64, HashMap<NodeIndex, NodeIndex>)> {
    let mut parent = HashMap::new();
    let mut path_cap = HashMap::new();
    let mut q = VecDeque::new();

    q.push_back(src);
    path_cap.insert(src, f64::INFINITY);

    while let Some(u) = q.pop_front() {
        for v in adj.get(&u)? {
            let residual = cap.get(&(u, *v)).copied().unwrap_or(0.0) - flow.get(&(u, *v)).copied().unwrap_or(0.0);
            if residual > 1e-9 && !path_cap.contains_key(v) {
                parent.insert(*v, u);
                path_cap.insert(*v, path_cap[&u].min(residual));
                if *v == sink {
                    return Some((path_cap[v], parent));
                }
                q.push_back(*v);
            }
        }
    }

    None
}